### ISSUE TITLE
"ThreadedDispatcherSignalTest.keeps_dispatching_after_signal_interruption" fails when mir_unit_tests run directly without filter

### ISSUE THREAD

**Issue by filman230** [MirServer/mir#1603]
I'm trying to build mir on archlinux using this pkgbuild   https://aur.archlinux.org/packages/mir-git/   but one test fails:

```
[ RUN      ] ThreadedDispatcherSignalTest.keeps_dispatching_after_signal_interruption
/home/user/mir-git/src/mir/tests/unit-tests/dispatch/test_threaded_dispatcher.cpp:391: Failure
Value of: result.exited_normally()
  Actual: false
Expected: true
[  FAILED  ] ThreadedDispatcherSignalTest.keeps_dispatching_after_signal_interruption (475 ms)
```

**Comment by wmww** [comment]
I'll boot up my Arch box and give it a try

**Comment by wmww** [comment]
The test passes when I build and run it normally but fails when I install from the AUR. Strange!

**Comment by wmww** [comment]
Hmm, it looks like there isn't a difference between a normal build and a build from the AUR:

- `make ptest` (which is supposed to run everything) succeeds on both
- `./bin/mir_unit_tests --gtest_filter='ThreadedDispatcherSignalTest.*'` succeeds on both
- Buuut `./bin/mir_unit_tests` seems to fail on both (even on Ubuntu!)

This appeared to be an Arch/AUR issue because it seems non of our other distros or CI run the tests in a way that triggers this problem. Changing issue title to reflect what I've found.

**Comment by AlanGriffiths** [comment]
On Tue, 14 Jul 2020 13:42:41 -0700, William Wold wrote:
> Hmm, it looks like there isn't a difference between a normal 
> build and a build from the AUR:
>
> - `make ptest` (which is supposed to run everything) succeeds on both
> - `./bin/mir_unit_tests 
> --gtest_filter='ThreadedDispatcherSignalTest.*'` succeeds on 
> both
> - Buuut `./bin/mir_unit_tests` seems to fail on both (even on Ubuntu!)

Some of our tests need an LD_PRELOAD of udevmock, ctest will use the right incantation.

>
> This appeared to be an Arch/AUR issue because it seems non of 
> our other distros or CI run the tests in a way that triggers 
> this problem. Changing issue title to reflect what I've found.
>

-- 
Sent using Dekko from my Ubuntu device

**Comment by wmww** [comment]
Should `ctest` be preferred to `make ptest` for automatic testing? I need to ask @vanyasem to change the PKGBUILD to one or the other.

**Comment by AlanGriffiths** [comment]
If the resources on the build system could be limited, then ptest might be too demanding. (I only mentioned ctest as a way to  run individual tests.) Why not `make test`?

**Comment by AlanGriffiths** [comment]
Ah, ok. Nothing in `mir_unit_tests` needs the preload. This looks like an unwanted interaction between tests. It looks as though `LogindConsoleServices` is modifying the process state:
```plain
mir_unit_tests --gtest_filter=LogindConsoleServices.*:ThreadedDispatcherSignalTest.*
mir_unit_tests --gtest_filter=-LogindConsoleServices.*
```

**Comment by AlanGriffiths** [comment]
Interestingly, this passes:
```plain
mir_unit_tests --gtest_filter=LogindConsoleServices.construction_fails_if_no_logind:ThreadedDispatcherSignalTest.*
```

**Comment by RAOF** [comment]
Ah. So:
```
bin/mir_unit_tests --gtest_filter=LogindConsoleServices.happy_path_succeeds:ThreadedDispatcherSignalTest.*
```
fails (the forked process in `ThreadedDispatcherSignaltest` dies with `SIGSEGV` on shutdown).

However
```
LD_PRELOAD=liblttng-ust-fork.so bin/mir_unit_tests --gtest_filter=LogindConsoleServices.happy_path_succeeds:ThreadedDispatcherSignalTest.*
``` 
Passes, as does `make test`.

This was fixed by #1496

**Comment by RAOF** [comment]
Ah! I see in your PKGBUILD you invoke `bin/mir_unit_tests` etc directly. You probably want to just run `make test` (or `make ptest`) as that'll pick up any more tests we add, and will avoid problems like this :grin:
